Created update_handlers.go

ssume1 opened the pull request:
Created a new file named update_handlers.go in the internal directory to manage user account update functionalities.

ssume1 committed:
Create update_handlers.go

Created a new file named update_handlers.go in the internal directory to manage user account update functionalities.

chrisdedman commented on backend/internal/handlers/update_handlers.go:
```diff
@@ -0,0 +1,97 @@
+    }
+
+    // Log out logic can be added here
+
```
Most of the logic for logging out a user is to revoke their token. We will need to add a new function in the token.go file that will revoke the current token (that should also be used for the logout feature). This could also be done in the front end, but we definitely want to expire a user token when login out on the backend too.

chrisdedman commented on backend/internal/handlers/update_handlers.go:
```diff
@@ -0,0 +1,97 @@
+        return c.Status(fiber.StatusNotFound).JSON(fiber.Map{"message": "User not found"})
+    }
+
+    hashedPassword, err := utils.HashPassword(password)
```
As mentioned on Discord, you will need to pass the new password as an object using the Entity.Account (similar to registration). However, because you may not be able to use the Entity.Account, you may want to create a new entity or find a way to pass the unchanged user name and email.

ssume1 committed:
Updated update_handlers.go

The UpdatePassword function in update_handlers.go updates the password, saves the new password in the database, and calls RevokeToken to revoke the user's token. Checks if the Authorization header is empty and handles cases where the token is provided with the "Bearer " prefix. Ensure that the token is properly revoked and the user is logged out after the password is updated.

ssume1 committed:
Update update_handlers.go

 Added the token to the update handler

ssume1 committed:
Delete backend/internal/handlers/update_handlers.go

ssume1 committed:
Added update_handlers.go

chrisdedman commented on update_handlers.go:
This file is in the wrong directory. 
It should be in ``backend/internal/handlers/``

chrisdedman commented on update_handlers.go:
```diff
@@ -0,0 +1,131 @@
+        return c.Status(fiber.StatusNotFound).JSON(fiber.Map{"message": "User not found"})
+    }
+
+    user.Username = username
```
nit: you could directly assign here the username, and remove line 22. You can do that here because the function here is pretty simple and easy to read/understand.

Example:
```go
user.Username = data["username"]
```

chrisdedman commented on update_handlers.go:
```diff
@@ -0,0 +1,131 @@
+    }
+
+    var user entity.Account
+    if err := database.DB.First(&user, userID).Error; err != nil {
```
nit: We may probably want to have this condition before checking if the email and confirmEmail are matching. This would avoid having unnecessary running conditions if the user does not exist in the database.
In other words, move lines 53 to 56 above ``if email != confirmEmail {``

chrisdedman commented on update_handlers.go:
```diff
@@ -0,0 +1,131 @@
+}
+
+// RevokeToken invalidates the token by setting its expiration time to the past
+func RevokeToken(tokenString string) error {
```
We probably want to move this to ``backend/internal/utils`` and call it here when needed with the ``utils.RevokeToken()``

chrisdedman commented on update_handlers.go:
```diff
@@ -0,0 +1,131 @@
+    userID := data["user_id"]
+    password := data["password"]
+    confirmPassword := data["confirm_password"]
+    tokenString := c.Get("Authorization")
```
Our token is stored in our Cookies in the frontend. We named it token. Something you can do is to fetch the token from the data like the above:

Example:
```go
tokenString := data["token"]
```

chrisdedman commented on update_handlers.go:
```diff
@@ -0,0 +1,131 @@
+        return c.Status(fiber.StatusNotFound).JSON(fiber.Map{"message": "User not found"})
+    }
+
+    hashedPassword, err := utils.HashPassword(password)
```
this will not work. You will need to pass the user reference to the HashPAssword.

Example:
```go
	if err := utils.HashPassword(&user); err != nil {
		return c.Status(fiber.StatusInternalServerError).JSON(fiber.Map{"message": "Couldn't hash password"})
	}
```

chrisdedman commented on update_handlers.go:
```diff
@@ -0,0 +1,131 @@
+        return c.Status(fiber.StatusInternalServerError).JSON(fiber.Map{"message": "Couldn't hash password"})
+    }
+
+    user.Password = hashedPassword
```
The problem here, this will not work because of how you handle the hashPassword above. A way to fix this, is to check if the user exists in the database (based on user ID), if it it does, then assign the new plain text password to the user, then call ``HashPassword(&user)``. Check if the password matches before assigning the new password to the user.Password.

Example:
```go
	var user entity.Account
	if err := database.DB.First(&user, userID).Error; err != nil {
		return c.Status(fiber.StatusNotFound).JSON(fiber.Map{"message": "User not found"})
	}

	if password != confirmPassword {
		return c.Status(fiber.StatusBadRequest).JSON(fiber.Map{"message": "Passwords do not match"})
	}

	user.Password = password

	if err := utils.HashPassword(&user); err != nil {
		return c.Status(fiber.StatusInternalServerError).JSON(fiber.Map{"message": "Couldn't hash password"})
	}
```

chrisdedman commented on update_handlers.go:
```diff
@@ -0,0 +1,131 @@
+    }
+
+    user.Password = hashedPassword
+    if err := database.DB.Save(&user).Error; err != nil {
```
This is good, but this should be done at the very end (before returning anything to the caller). For example, if the RevokeToken fails, we do not want to have the updated password stored in the database. We want to save the changes when everything is passing and we get no error. Consider moving this condition above the function's last return statement.

chrisdedman commented on update_handlers.go:
```diff
@@ -0,0 +1,131 @@
+
+    // Validate and revoke the token after password change
+    if tokenString == "" {
+        return c.Status(fiber.StatusUnauthorized).JSON(fiber.Map{"message": "Authorization header is missing"})
```
nit: change the Authorization header to the token.

Example:
```go
return c.Status(fiber.StatusUnauthorized).JSON(fiber.Map{"message": "Token is missing"})
```

chrisdedman commented on update_handlers.go:
```diff
@@ -0,0 +1,131 @@
+    }
+
+    // If the token is in the format "Bearer <token>", remove "Bearer "
+    if len(tokenString) > 7 && tokenString[:7] == "Bearer " {
```
This should not be needed because of the way we handle our token, but I guess it is okay to leave it there in case we update the way we handle our authorization.

chrisdedman reviewed (changes requested):
You have some mistakes in your code. Especially because this would not compile (the file is placed in the wrong directory, the flow of the code is in the wrong order, wrong uses of functions, etc). I would recommend testing any changes before pushing anything, like this it would avoid having broken changes in review.

Also, when writing code, think about the flow of your algorithm. You should think about it like a logical story that you are trying to tell. For example, if you want to update some information from a user. You want to validate first if the user exists in the database, if it does not, you do not need to go any further because any steps will be useless, thus return an error message. If it exists, you want to move to the next step. If you are trying to update the password, you want to have some data validation like making sure that the password matches the password confirmation. If it does not, no need to go further, you can stop here and return an error message, if it does match, you can proceed to the next step which is to hash the password, etc. 

If this is not clear, please feel free to ask me on Discord to clarify.

ssume1 committed:
Fixed update_handlers.go
